ls1intum/Artemis
# `Plagiarism checks`: Introduce min length parameter for programming exercises

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises
_This change is not connected to CI/CD configuration._

### Motivation and Context
Artemis supports the minimum submission size parameter for plagiarism checks of text and modeling exercises but not for programming exercises. This parameter is very useful for removing false positives form the comparison.

### Description
Minimum size parameter for programming exercises plagiarism checks is introduced as minimal number of lines changed, according to `git diff`, between the template and the solution.

This PR also improves the UX by introducing minimum size tooltips dedicated to each exercise type.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise

1. Create two identical submissions that change the same file in the programming exercises template.
2. Run plagiarism detection on those submissions without minimum size set.
3. Verify that the results is present.
4. Set the minimum size to a number larger than the changed lines (ex. 50) and re-run detection.
5. Verify that your submissions are not included in the results or, if those were the only ones, that a warning was shown.

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| PlagiarismDetectionService.java | 100% | ✅                           |
| plagiarism-inspector.component.ts | 88% | ✅                           |


### Screenshots
Minimum size option for programming exercises:
<img width="1228" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/22077541-1163-4ef8-a6b9-d318ad0f1dd6">
